### ISSUE TITLE
Trying to get avahi to work: Daemon not running

### ISSUE THREAD

**Issue by pcwalden** [dnschneid/crouton#2331]
I am new to chromebooks and crouton, but not linux. I cannot find any issue/wiki item to help me with getting avahi running in the crouton xfce4-desktop(Ubuntu 12.04.5 precise), so that my ASUS C300M can access other local domain services on my LAN.

I am able to start the avahi-daemon with "sudo start avahi". I can also see the daemon running using pgrep. However, when I run avahi-resolve-host-name <xxx>.local, I get the error "Failed to create client object: Daemon not running". I checked the /etc/avahi/avahi-daemon.conf against my other Fedora and Debian machines and it looks the same.

I am thinking that the firewalld is blocking the 5353 port that avahi uses. However, I do not know how to check the firewalld settings on a chromebook or change them. Google search does not seem to answer anything either. The firewalld man page is no help as the configuration files in /etc/firewalld do not seem to exist either. I suspect Google has hidden them somewhere.

I did try something I saw in the wiki entry "Running servers in crouton". I installed iptables, then ran manually "sudo iptables -A INPUT -p udp -m udp --dport 5353 -j ACCEPT". I then restarted the avahi-daemon, but I still get the error.

Any help would be appreciated.

**Comment by pcwalden** [comment]
I ran an iptables listing and the changes I added are indeed there. There also seems to be a Apple iTunes entry as well. Still getting the "Failed to create client object: Daemon not running" error.

>(precise)pwalden@localhost:~$ sudo iptables -L
>[sudo] password for pwalden: 
>Chain INPUT (policy DROP)
>target     prot opt source               destination         
>ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
>ACCEPT     all  --  anywhere             anywhere            
>ACCEPT     icmp --  anywhere             anywhere            
>ACCEPT     udp  --  anywhere             224.0.0.251          udp dpt:mdns
>NFQUEUE    udp  --  anywhere             anywhere             NFQUEUE num 10000
>ACCEPT     udp  --  anywhere             anywhere             udp dpt:mdns
>
>Chain FORWARD (policy DROP)
>target     prot opt source               destination         
>
>Chain OUTPUT (policy DROP)
>target     prot opt source               destination         
>NFQUEUE    udp  --  anywhere             239.255.255.250      udp dpt:1900 NFQUEUE num 10001
>ACCEPT     all  --  anywhere             anywhere             ctstate NEW,RELATED,ESTABLISHED
>ACCEPT     all  --  anywhere             anywhere            

I installed mdns-scan and it successfully found all my local LAN avahi services, but as its man page says, it does not use the avahi-daemon. It only shows that they are accessible.

**Comment by pcwalden** [comment]
Here is a grep of the avahi-daemon records in the /var/log/message file:

[avahi-daemon_message_log.txt](https://github.com/dnschneid/crouton/files/75660/avahi-daemon_message_log.txt)

In the log it says "WARNING: Detected another IPv4 mDNS stack running". I assume this is the chrome OS stack and the ubuntu stack? Is this a source of the problem?

The log also says "WARNING: No NSS support for mDNS detected, consider installing nss-mdns", but the package manage says libnss-mdns is already installed. The /etc/nsswitch.conf file "hosts" record seems correct as well.

>hosts:          files mdns4_minimal [NOTFOUND=return] dns mdns4

**Comment by pcwalden** [comment]
When I Shift+Ctrl+Alt+< back to the Chrome environment, I do see a different /etc/nsswitch.conf file with just the following:

>hosts: files dns

So I guess this is why the avahi-daemon is complaining about nss-mdns not being installed.

How would crouton allow access to the Chrome /etc/nsswitch.conf? I see that the Downloads folder is mapped back to the Chrome user's Downloads folder.

**Comment by pcwalden** [comment]
I tried making changes to the /etc/avahi/avahi-daemon.conf file by setting "host-name=chromebook" and I noticed that the avahi-daemon still tried to register the chromebook's native hostname. I suspected the daemon is looking in the real "/etc" and not finding the conf file. So I re-ran  avahi-daemon as a non-daemon and lo and behold mdns and other LAN services are working! It also stopped complaining about installing nss-mdns. So as a daemon, it is not finding the  avahi-daemon.conf file. It could also be failing because it is accessing the wrong nsswitch.conf file as well.

I will try to come up with a simple recipe to get it working and post it to the wiki.

**Comment by pcwalden** [comment]
Added the recipe to the existing [Running Servers](https://github.com/dnschneid/crouton/wiki/Running-servers-in-crouton) page.

If someone wants to test the recipe, we can close this issue.

**Comment by pcwalden** [comment]
So far the only problem I have found is when I log out of my xfce4-desktop chroot and try to backup the chroot. It edit-chroot complains that it cannot unmount the chroot as it is still being used. If I pkill the avahi-daemon, then the `edit-chroot -b` proceeds, but still asks to force unmount the chroot as something is still using it. I will try to check out the problem further.

**Comment by shelleyvadams** [comment]
**TL;DR**
Chrome OS is running its own avahi-daemon, with `host-name-from-machine-id=yes` in `/etc/avahi/avahi-daemon.conf`

***

I followed your instructions in the wiki.

The good news:
Discovery worked fine.

The bad news:
I ran into problems when I tried to set up a service (for apache) on my chromebook.

I un-commented `host-name=foo` in `/etc/avahi/avahi-daemon.conf` and set my own hostname -- e.g., `host-name=chromewebdev` -- but instead of the expected hostname of `chromewebdev.local` my IP address resolved to `b499d7ec45600bef38740391574210c4.local`. 

Things got weirder when I went to kill the avahi-daemon process. Not yet familiar with pkill and pgrep (thanks for the tip there) I did `ps -Af | grep avahi`. There were four lines for avahi-daemon: 
* two `avahi-daemon: chroot helper`
* one `avahi-daemon: running [b499d7ec45600bef38740391574210c4.local]`
* one `avahi-daemon: registering [chromewebdev-5.local]`

When I killed avahi-daemon, two of those processes refused to die, and the were still there after unmounting the crouton chroot. That made me suspect these processes actually belonged to Chrome OS. Which would also explain your
> WARNING: Detected another IPv4 mDNS stack running

To find out, I rebooted tried `ps -Af | grep avahi` without ever having started the crouton chroot. Sure enough, I had two avahi-daemon processes:
* one `avahi-daemon: running [a1d00c7453b1f13024d9e9be57423f26.local]`
* one `avahi-daemon: chroot helper`

The hostname changed, what the ⁉️ 

So... what's up with the hostname in Chrome OS `/etc/avahi/avahi-daemon.conf`?
A six-line explanatory comment followed by `host-name-from-machine-id=yes`
Meanwhile the Chrome OS `/etc/avahi/services/` directory is empty.



***
Chrome OS:
Version 51.0.2704.55 beta (64-bit)
Platform 8172.39.0 (Official Build) beta-channel gnawty
Firmware Google_Gnawty.5216.239.34

crouton: version 1-20160518095657~master:fd37bb79
release: jessie

**Comment by shelleyvadams** [comment]
A couple additional notes:

It's possible to change the Chrome OS avahi hostname by using `avahi-set-host-name` from a crosh shell; e.g. `chronos@localhost / $ sudo avahi-set-host-name -v chromewebdev`. Obviously, it's not persistent, but neither is Chrome's default 32-character hex string.

Setting `disable-publishing=yes` in `/etc/avahi/avahi-daemon.conf` allows for service discovery in a crouton chroot and prevents the daemon from repeatedly trying (and failing) to register its hostname.
